### DISCUSSION TITLE
Multi Display Issues with App Layouts

### DISCUSSION THREAD

**Discussion by nblopes** [rxhanson/RectanglePro-Community#535] [General]
Equipment:
MacBook Pro M3 w/ macOS Sonoma 14.5
Samsung G95NC, Dual Display Setup (works as two display side by side)

Scenario:
MacBook Pro in Clam shell mode (lid closed) connected to the two displays
An App Layout is set for the displays on the G9 where each window has snap area ON.

Expected result:
When disconnecting the MacBook Pro and opening it.
Was expecting that App Layouts with its Snap Areas to stick to the display/monitor used to create it (Moom does this).

Problem:
Rectangle Pro does not. The MacBook Pro display assumes the App Layout and snap areas set to another display, a bigger and wider display. So all snap areas are invalid as it sizes and position do not make sense in the context of the MacBook Pro Display (don't work).

This makes it impossible to use this app as I am frequently connecting and disconnecting my laptop and using it without being connected external display(s).

**Comment by rxhanson** [comment]
Sorry it's not working for you. Can you send me your configuration and I can see what's going on? You can export it from the general tab of the settings window.

By default app layouts are created using absolute pixel positions in the coordinate space, so perhaps a potential scenario here is that connecting displays re-adjusts the coordinate space and creates a discrepancy, depending on your setup.

**Comment by nblopes** [comment]
Yes I think that is the problem. I was under impression that App Layouts and its snap areas would stick to the monitor in which they were defined.

Here is demo showing the impact of not doing so ...

https://youtu.be/ZEtmmnYuHRY

**Reply by rxhanson** [reply]
Thanks for the video! Indeed, the issue is the coordinate space adjustment. As an imperfect workaround until I can get something out to fix or improve this, try changing your saved layouts to be relative values instead of absolute. To do this, you can change the W value to 1/2 and H value to 1 for all of your layouts (sort of tedious, I know). You might have to add another layout for the MacBook Pro Display, and set the Display values in the layout as well.

**Reply by nblopes** [reply]
I will try and provide you with feedback. 

But wouldn't be a better solution to have snap areas within layouts stick to the display they were created in? Or even allow after the fact to define the display where a given snap area within a layout should be active and no other?

**Reply by jqning** [reply]
> changing your saved layouts to be relative values instead of absolute

Is there an instruction for how to do this? I usually run tethered to a screen, but I am setting up OBS so I want my Window layout to be associated with the screen that will always be attached, which is the laptop screen. How do I make it so Rectangle Pro always sends the apps for "layout x" to my laptop screen?

**Reply by rxhanson** [reply]
Currently, Rectangle Pro is only concerned with locations of displays in a 2D plane, and in macOS that 2D plane is positioned around your "main" display having an origin of (0,0). This means that if your laptop screen is set as the "main" display in System Settings regardless of connected displays, then it will always keep those coordinates and absolute positioning would be desired in this scenario. 

There are obvious limitations with this approach, and I have plans to adjust the layout mechanism in Rectangle Pro to allow for scenarios like what you're looking for, even if your laptop display is not set as the main display when you have multiple displays connected.

Changing saved layouts to relative values would mean manually taking the sizes/positions you want and giving them a docking position and w/h values between 0 and 1 (percentages of the screen). For example, say you have a saved layout for safari that's in the left half of your display, then you could change it to a docking position of left and w:0.5, h:1. There's no way to automatically do this, and I would still like to provide this in a future update.
